### PR TITLE
esp32/ota: Implement ESP-IDF OTA functionality.

### PR DESCRIPTION
**Pull request by ekondayan**
UPDATE: The test completed successfully on NodeMCU ESP32 by ai-thinker with ESP-IDF v4.0, v4.1, v4.2, v4.3

Implemented new functions in esp32.Partition:
* mark_app_invalid_rollback_and_reboot()
* check_rollback_is_possible()
* app_description()
* app_state()
* ota_begin()
* ota_write()
* ota_write_with_offset() for ESP-IDF version >= 4.2
* ota_end()
* ota_abort() for ESP-IDF version >= 4.3

* create tests
* update documentation

For many commercial products, Over The Air updates are a very important and critical part. It must be reliable and should not brick the device. Writing a good OTA from scratch is a daunting task. For that reason the use of well tested and proven reliable libraries is much more preferable than the ones developed in the house.

USECASE
=======
I'm developing an industrial device where the OTA is an essential part of it. Since only a few functions from esp-idf are 
implemented (enough for hobby project but not enough for commercial project), I ended up duplicating the esp-idf ota functionality in python. 
The result was a module with a questionable quality. I tried to predict all the possible places where it could crash, but my gut was 
telling me that I could be missing something. So I decided to implement more of the the OTA functions from esp-idf and to use them in my OTA module.
I've rewritten my OTA module and replaced the redundant code with the implemented functions from esp-idf. This allowed me to reduced the size of the module significantly, increase the robustness of the code and on top of that now the code got much simpler and easier to maintain.
The total increase in size of the compiled app image is 2432 bytes. 

IMPLEMENTATION
==============
  Extend the esp32.Partition class where all the OTA related functions are 
prefixed with "ota_" and app related functions are  prefixed with "app_".

Example:
  from esp32 import Partition
  app_part = Partition(Partition.RUNNING)
  app_part.app_description()
  app_part.app_state()
  handle = app_part.ota_begin()
  app_part.ota_end(handle)

New functions:
  Partition.mark_app_invalid_rollback_and_reboot(cls)
  Partition.check_rollback_is_possible(cls)
  Partition.app_description(self)
  Partition.app_state(self)
  Partition.ota_begin(self, image_size = 0)
  Partition.ota_write(self, handle_in, data_in)
  Partition.ota_write_with_offset(self, handle_in, data_in, offset)
  Partition.ota_end(self, handle_in)
  Partition.ota_abort(self, handle_in) only for ESP-IDF version >= 4.3

BENEFITS
========
* no code duplication
* reduced code size
* cleaner and easier to maintain code
* esp-idf handles encrypted flash
* use a reliable and well tested library written by the creators of ESP32
* more robust OTA procedure
* better performance

CONS
====
* None :) (maybe the 2432 bytes overhead in the final compiled app image)

### CODE REVIEW

**Comment by ekondayan** [comment]
Anybody? Am I the only one who find this useful?

**Comment by beyonlo** [comment]
Hi! Nice work!

Is this OTA feature in the 1.17 version, or was committed in the mainstream to tests?

If not, are there a roadmap to have an OTA by on ESP32 MicroPython?

I would like to test it!

Thank you.

**Comment by ekondayan** [comment]
Hi,
unfortunately it is still not implemented mainstream. We are the only two people here who find this useful. The PR was submitted 6 months ago and there was not even a single comment from the community for that time.  

:(

You can find an example in the test section: tests/esp32/partition_ota.py
You can find the documentation here: docs/library/esp32.rst
To test it, just copy the tests/esp32/partition_ota.py to your freshly flashed esp32 and run it.

**Comment by kevinkk525** [comment]
I'd find an OTA option very useful too and have been following this PR.

**Comment by beyonlo** [comment]
@dpgeorge and/or @jimmo or someone else.

Please, could you do a bit attention for this important PR?

1. In my opinion OTA feature is a very useful feature. In real real life, in the pratical way, is an essential feature, because will be possible to update easily the MicroPython version and applications via WiFi on the ESP32, for example. 
2. Another great thing in this PR is this OTA implementation following the same and tested (I hope) OTA from ESP-IDF implementation - this is great for evolution of OTA on MicroPython, following always from ESP-IDF.

Could you please to check, review or suggest any changes on this PR? And if necessary, the @ekondayan can to do the necessary changes to this amazing feature enter in the mainstream, and maybe ready to MicroPython 1.18.

Thank you so much for your attention.

**Comment by lovetheguitar** [comment]
We are also following this PR very closely. Would make lots of things a lot more maintainable here. 

Thanks for all your work @ekondayan! It is greatly appreciated. 😺

**Comment by beyonlo** [comment]
@ekondayan Hello.

This PR will works just on ESP32, or will works in all Espressif chips that use the ESP-IDF, like as ESP32-S2, ESP32-S3, ESP32-C3 and ESP32-C6?

As I know, as this PR use the ESP-IDF OTA, so I think that this PR will works in all that chips, Am I right?

Did you tested it in other chip than ESP32?

Thank you

**Comment by ekondayan** [comment]
Yes this is just a wrapper around the esp-idf. It should work on all chips supported by esp-idf.

**Comment by beyonlo** [comment]
Hi @ekondayan 

> Yes this is just a wrapper around the esp-idf. It should work on all chips supported by esp-idf.

Just to clarify: this wrapper is around this ESP OTA? -> https://github.com/espressif/esp-idf/tree/master/examples/system/ota

If not, what this ESP OTA differ from another ESP OTA that you has wrapped?

Thank you in advance!

**Comment by ekondayan** [comment]
I suppose that you refer to "ESP HTTPS OTA", which itself uses a lower level API.

This PR is about the lower level API, which can be found here -> [https://github.com/espressif/esp-idf/blob/ef98a36/components/app_update/include/esp_ota_ops.h](https://github.com/espressif/esp-idf/blob/ef98a36/components/app_update/include/esp_ota_ops.h)

Link to the documentation -> [https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/ota.html](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/ota.html)

> If not, what this ESP OTA differ from another ESP OTA that you has wrapped?

This is the only OTA I have wrapped so far :)

**Comment by demetz** [comment]
Hi,

do you have a small sample project demonstrating the functionality?

**Comment by ekondayan** [comment]
@demetz Have a look at the test file in this PR

**Comment by beyonlo** [comment]
Hello @ekondayan  

> You can find an example in the test section: tests/esp32/partition_ota.py 
> To test it, just copy the tests/esp32/partition_ota.py to your freshly flashed esp32 and run it.

I did that but do not works, show some errors. I'm using ESP32 with OTA firmware https://micropython.org/download/esp32-ota/ 

Here the errors:

```
>>> import os
>>> os.uname()
(sysname='esp32', nodename='esp32', release='1.18.0', version='v1.18 on 2022-01-17', machine='4MB/OTA module with ESP32')
>>> import partition_ota
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "partition_ota.py", line 143, in <module>
AttributeError: 'Partition' object has no attribute 'app_state'
>>>
```

Well, just to a try, I commented that lines, and have other errors:

```
>>> import partition_ota
Partition copy: (0, 16, 65536, 1572864, 'ota_0', False) --> (0, 16, 65536, 1572864, 'ota_0', False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "partition_ota.py", line 159, in <module>
  File "partition_ota.py", line 110, in copy_partition_incr
AttributeError: 'Partition' object has no attribute 'ota_begin'
>>> 
```

Am I doing something wrong?

Thank you.

**Comment by ekondayan** [comment]
> Am I doing something wrong?
  Yes. 

The image you are using is not based on my pull request. To use the OTA functions that I've implemented in my pull request, you have to check out the branch `feature/ota` from my fork and then to compile the image yourself.

**Comment by beyonlo** [comment]
Hello @ekondayan 

> The image you are using is not based on my pull request. To use the OTA functions that I've implemented in my pull request, you have to check out the branch `feature/ota` from my fork and then to compile the image yourself.

I used that branch but the results was `SKIP`. It looks like I'm not compiling your branch, but it says it's the correct branch (`feature/ota`) on the git below. Any idea what is the error?

```
pi@pi:~/esp/esp-idf$ git clone -b feature/ota --single-branch https://github.com/ekondayan/micropython.git
pi@pi:~/esp/esp-idf$ cd micropython/
pi@pi:~/esp/esp-idf/micropython$ git branch -a 
* feature/ota
  remotes/origin/feature/ota
pi@pi:~/esp/esp-idf/micropython$ make -C mpy-cross
...
pi@pi:~/esp/esp-idf/micropython$ cd ports/esp32
pi@pi:~/esp/esp-idf/micropython/ports/esp32$ make submodules
git submodule update --init ../../lib/berkeley-db-1.xx
Submodule 'lib/berkeley-db-1.xx' (https://github.com/pfalcon/berkeley-db-1.xx) registered for path '../../lib/berkeley-db-1.xx'
Cloning into '/home/pi/esp/esp-idf/micropython/lib/berkeley-db-1.xx'...
Submodule path '../../lib/berkeley-db-1.xx': checked out '35aaec4418ad78628a3b935885dd189d41ce779b'
pi@pi:~/esp/esp-idf/micropython/ports/esp32$ make
idf.py -D MICROPY_BOARD=GENERIC -B build-GENERIC  build
...

pi@pi:~/esp/esp-idf/micropython/ports/esp32$ /home/pi/.espressif/python_env/idf4.4_py3.8_env/bin/python ../../../components/esptool_py/esptool/esptool.py -p /dev/ttyUSB0 -b 460800 --before default_reset --after hard_reset --chip esp32  write_flash --flash_mode dio --flash_size detect --flash_freq 40m 0x1000 build-GENERIC/bootloader/bootloader.bin 0x8000 build-GENERIC/partition_table/partition-table.bin 0x10000 build-GENERIC/micropython.bin


>>> os.uname()
(sysname='esp32', nodename='esp32', release='1.18.0', version='v1.14-1291-gefc889cf8 on 2022-05-27', machine='ESP32 module with ESP32')
>>> import partition_ota
SKIP
>>> 
```

Thank you in advance!

**Comment by beyonlo** [comment]
Any idea about that @ekondayan?

Thank you.

> Hello @ekondayan
> 
> > The image you are using is not based on my pull request. To use the OTA functions that I've implemented in my pull request, you have to check out the branch `feature/ota` from my fork and then to compile the image yourself.
> 
> I used that branch but the results was `SKIP`. It looks like I'm not compiling your branch, but it says it's the correct branch (`feature/ota`) on the git below. Any idea what is the error?
> 
> ```
> pi@pi:~/esp/esp-idf$ git clone -b feature/ota --single-branch https://github.com/ekondayan/micropython.git
> pi@pi:~/esp/esp-idf$ cd micropython/
> pi@pi:~/esp/esp-idf/micropython$ git branch -a 
> * feature/ota
>   remotes/origin/feature/ota
> pi@pi:~/esp/esp-idf/micropython$ make -C mpy-cross
> ...
> pi@pi:~/esp/esp-idf/micropython$ cd ports/esp32
> pi@pi:~/esp/esp-idf/micropython/ports/esp32$ make submodules
> git submodule update --init ../../lib/berkeley-db-1.xx
> Submodule 'lib/berkeley-db-1.xx' (https://github.com/pfalcon/berkeley-db-1.xx) registered for path '../../lib/berkeley-db-1.xx'
> Cloning into '/home/pi/esp/esp-idf/micropython/lib/berkeley-db-1.xx'...
> Submodule path '../../lib/berkeley-db-1.xx': checked out '35aaec4418ad78628a3b935885dd189d41ce779b'
> pi@pi:~/esp/esp-idf/micropython/ports/esp32$ make
> idf.py -D MICROPY_BOARD=GENERIC -B build-GENERIC  build
> ...
> 
> pi@pi:~/esp/esp-idf/micropython/ports/esp32$ /home/pi/.espressif/python_env/idf4.4_py3.8_env/bin/python ../../../components/esptool_py/esptool/esptool.py -p /dev/ttyUSB0 -b 460800 --before default_reset --after hard_reset --chip esp32  write_flash --flash_mode dio --flash_size detect --flash_freq 40m 0x1000 build-GENERIC/bootloader/bootloader.bin 0x8000 build-GENERIC/partition_table/partition-table.bin 0x10000 build-GENERIC/micropython.bin
> 
> 
> >>> os.uname()
> (sysname='esp32', nodename='esp32', release='1.18.0', version='v1.14-1291-gefc889cf8 on 2022-05-27', machine='ESP32 module with ESP32')
> >>> import partition_ota
> SKIP
> >>> 
> ```
> 
> Thank you in advance!

**Comment by ekondayan** [comment]
Sorry for the delay.

It looks like your current partition name does not starts with `ota_`.
For the test to work, you have to have two partitions of subtype ota_1 and ota_2 named ota_1 and ota_2.
Take a look at `micropython/ports/esp32/partitions-16MiB-ota.csv`

More info:
[](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/partition-tables.html)

**Comment by ekondayan** [comment]
Another requirement is to enable app rollback in the sdkconfig file

`CONFIG_BOOTLOADER_APP_ROLLBACK_ENABLE=y`

**Comment by w0ssh** [comment]
cvxzbnhgfd@gmail.com
لقد اخترقو حسابي و قامو بتغير كلشي

**Comment by abraha2d** [comment]
@ekondayan Thanks for keeping this up to date, I'm using this successfully in this project. Hopefully it can get mainlined soon.

**Comment by ekondayan** [comment]
@abraha2d Thank you for your positive feedback. I am glad that it is useful to you. 

@dpgeorge I've been keeping this PR up do date for almost 2 years. Is there a slight chance this PR to go mainline?

**Comment by glenn20** [comment]
> Another requirement is to enable app rollback in the sdkconfig file
> 
> `CONFIG_BOOTLOADER_APP_ROLLBACK_ENABLE=y`

This is now enabled in the v1.21.0 release. See #12475 (sorry - I missed this PR when I was preparing that PR).

Note - if you are interested in OTA, I also have a micropython OTA tool at <https://github.com/glenn20/micropython-esp32-ota> which uses the existing support for OTA in micropython. There is also a separate tool at <https://github.com/glenn20/mp-image-tool-esp32> which can (among other things) add OTA partition tables to micropython firmware images and flash storage on ESP32 devices.

**Comment by projectgus** [comment]
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.

File: docs/library/esp32.rst [hunk 1]
```diff
@@ -173,15 +173,154 @@ methods to enable over-the-air (OTA) updates.
...
+   
+    If the object does not contain an app partition, OsError exception will be raised:
+    ``ESP_ERR_NOT_FOUND`` no app description structure is found. Magic word is incorrect.
+    ``ESP_ERR_NOT_SUPPORTED`` Partition is not application.
+    ``ESP_ERR_INVALID_ARG`` Partition’s offset exceeds partition size.
+    ``ESP_ERR_INVALID_SIZE`` Read would go out of bounds of the partition.
+
+.. method:: Partition.app_state()
+
+    Returns the app state of a valid ota partition. It can be one of the following strings:
```
Comment (projectgus) [inline docs/library/esp32.rst]: The more common pattern in MicroPython would be to return an integer from this function, and define these as constants on the Partition class - i.e. `Partition.APP_STATE_NEW`, `Partition.APP_STATE_VERIFY`, etc.

File: docs/library/esp32.rst [hunk 2]
```diff
@@ -173,15 +173,154 @@ methods to enable over-the-air (OTA) updates.
...
+.. classmethod:: Partition.mark_app_invalid_rollback_and_reboot()
+
+    Mark the current app partition invalid by writing to the "otadata"
+    partition, rollback to the previous workable app and then reboots.
+    If the rollback is sucessfull, the device will reset.  If the flash does not have 
+    at least one valid app (except the running app) then rollback is not possible.
+    If the "CONFIG_BOOTLOADER_APP_ROLLBACK_ENABLE" option is set, and a reset occurs without
+    calling either 
+    ``mark_app_valid_cancel_rollback()`` or ``mark_app_invalid_rollback_and_reboot()``
+    function then the application is rolled back.
```
Comment (projectgus) [inline docs/library/esp32.rst:194]: Is there a difference between calling this API, versus simply restarting without calling `mark_app_valid_cancel_rollback()`? If we can get the same behaviour without adding a new function here then it might be better to leave it out?

**Review by projectgus — COMMENTED** [review]
Thanks for keeping this PR up to date for so long, @ekondayan.

The state of OTA support in MicroPython has probably changed since you first opened the PR, but if I follow correctly MicroPython already has implemented OTA functionality for ESP32 and this PR adds three related areas of functionality:

1. New APIs for reading app information from the partition.
2. Additional rollback support for marking an update invalid or testing if rollback is possible.
3. A wrapper around the ESP-IDF OTA write API rather than writing to the partition directly.

Part 1 makes sense to me as useful functionality for managing OTA updates in a deployed application. You could probably put these into a separate PR and we could merge them pretty quickly.

Part 2 I have some inline questions about, but the `check_rollback_is_possible()` function definitely seems useful to avoid reboot loops.

Part 3, I'm not certain what the exact benefits of an `ota_*` API is here compared to manually writing the partition and then calling `part.set_boot()` to set it as the next boot partition.

Is the main difference that the OTA API will verify the image? If so, could we add this as an optional boolean `verify` argument on `set_boot()` and get the same functionality? If there are any other benefits, can you please explain them?

Finally, the code size report suggests this PR adds 1.6KB of static RAM usage (`+1632(data) +8(bss)]`). I'm not sure how or why, there's no static buffer that I can see, but that's a big impact for ESP32s without PSRAM so it'd be good if there was a way to avoid it.

File: docs/library/esp32.rst (hunk 1 above)
Comment (ekondayan) [reply docs/library/esp32.rst]: You're right, that aligns better with MicroPython conventions. I'll refactor app_state() to return the integer value directly and add the corresponding constants to the Partition class:
```
Partition.APP_STATE_NEW
Partition.APP_STATE_PENDING_VERIFY  
Partition.APP_STATE_VALID
Partition.APP_STATE_INVALID
Partition.APP_STATE_ABORTED
Partition.APP_STATE_UNDEFINED
```

File: docs/library/esp32.rst (hunk 2 above)
Comment (ekondayan) [reply docs/library/esp32.rst:194]: Yes, there is a meaningful difference. When you simply restart without calling `mark_app_valid_cancel_rollback()`, the rollback only happens on the next boot (the bootloader detects the "pending_verify" state persisted across two boots and then changes it to "aborted") - so it requires two boot cycles.
In contrast, `mark_app_invalid_rollback_and_reboot()` performs an immediate rollback in one boot cycle - it actively marks the current partition as invalid in otadata and reboots to the previous working partition in one atomic operation.

1. Post-validation failures: If the app already called `mark_app_valid_cancel_rollback()` early in boot and later detects a critical issue (can't reach server, hardware fails under load, config corruption), passive rollback is no longer possible - the partition is already marked valid.
2. Atomicity: A manual approach risks power loss between operations, leaving undefined state. The ESP-IDF function writes otadata and reboots as a single coordinated operation.

**Comment by ekondayan** [comment]
Benefits of an `ota_*` is that they handle efuse values, checksums, encryption, etc. and they follow the best practices for manipulation the ota partition. These are mission critical functions and for a production devices I wouldn't trust myself reimplementing them in micropython. 

TL;DR
ESP-IDF provide fast, small and reliable functions that cover all edge cases and follow best practices for esp32 devices. 

1. Image validation: `ota_end() `validates the complete image (magic bytes, checksums). With direct writes, you'd need to implement validation yourself.
2. Secure boot signature verification: If secure boot is enabled, `ota_end()` verifies the firmware signature. This is non-trivial to replicate manually.
3. Flash encryption handling: The OTA API handles encrypted flash transparently. Direct writes to an encrypted partition require manual encryption handling.
4. Rollback state initialization: The OTA API properly sets the new partition's state to "new" in otadata, which triggers the rollback state machine on next boot. Direct writes + `set_boot()` might not initialize this correctly.
5. Anti-rollback/secure version: If anti-rollback is enabled, the OTA API validates that the new firmware's secure_version >= efuse value.
6. Out-of-order writes: `ota_write_with_offset()` allows non-contiguous writes when network packets arrive out of order - useful for UDP-based OTA.
7. Clean abort: `ota_abort()` properly cleans up state if the update fails mid-way.
8. verify on `set_boot()` isn't sufficient: Validation needs to happen after all writes complete but before marking bootable. The handle-based API (ota_begin → ota_write → ota_end) provides this atomic commit-or-abort semantic. By the time you call `set_boot()`, you've already written potentially invalid data.

**Comment by projectgus** [comment]
Hi @ekondayan,

Thanks for getting back to me.

> * Image validation: `ota_end() `validates the complete image (magic bytes, checksums). With direct writes, you'd need to implement validation yourself.
> * Secure boot signature verification: If secure boot is enabled, `ota_end()` verifies the firmware signature. This is non-trivial to replicate manually.

There's a function [esp_image_verify()](https://github.com/espressif/esp-idf/blob/master/components/bootloader_support/include/esp_image_format.h#L114) which we can call directly (it's what `esp_ota_end()` is calling internally anyhow).

> * Flash encryption handling: The OTA API handles encrypted flash transparently. Direct writes to an encrypted partition require manual encryption handling.

The ESP-IDF OTA API is calling the same `esp_partition_write()` functions that MicroPython already calls, so this part is the same.

---

Before I keep answering these points, I'm sorry but I have to ask an important question: did you use generative AI (like ChatGPT) to respond to my review? I don't mean to be rude, but many of the answers have the "confident and mostly but not totally correct" style of something written by an LLM. I need to know that you're putting in time and effort into creating these comments, before I put more of my own time and effort into responding to them.

**Comment by ekondayan** [comment]
> Before I keep answering these points, I'm sorry but I have to ask an important question: did you use generative AI (like ChatGPT) to respond to my review? I don't mean to be rude, but many of the answers have the "confident and mostly but not totally correct" style of something written by an LLM. I need to know that you're putting in time and effort into creating these comments, before I put more of my own time and effort into responding to them.

Of course I use LLM. It would be weird if I don't. You are not rude at all to question this. I myself avoid investing effort and energy unless I see a reasonable and thinking AI or NI(Natural Intelligence) on the other end. I don't use LLM to generate my response, just to review and polish. The technical reasoning is mine.

> There's a function [esp_image_verify()](https://github.com/espressif/esp-idf/blob/master/components/bootloader_support/include/esp_image_format.h#L114) which we can call directly (it's what esp_ota_end() is calling internally anyhow).

You're right that `esp_image_verify()` exists and the same write functions are used internally but my concern is about the complete workflow.

I see where this leads - you want to save few bytes and rely on the developers to implement the correct sequence. My argument for that is: for OTA enabled devices, this is the one of the most important features, because if not implemented correctly, you can render the device useless and end its life prematurely. Imagine if you sell or deploy thousands of devices which can get bricked easily by a tiny human mistake. It depends on who you expect to use MicroPython. If you are targeting DIY hobbyists who primarily use it for watering their plants, you can take the risk and rely on the developers to implement the correct workflow manually, but if you want to ship commercial devices that sit in factories where downtime is unacceptable, you can not rely solely on devs to implement it correctly. I wouldn't trust even myself for that. It is too risky to deviate from the official battle tested workflow recommended by Espressiff.

I can make this feature optional via a flag in `mpconfigport.h`. Users who need the full OTA workflow can just enable it. 
``` c
#ifndef MICROPY_PY_ESP32_PARTITION_OTA_EXTENDED
#define MICROPY_PY_ESP32_PARTITION_OTA_EXTENDED (1)
#endif
```
What do you think?

TL;DR
Manually calling `write` -> `verify` -> `set_boot` is error-prone and requires reading the IDF documentation carefully. On the other hand ESP-IDF OTA API enforces the correct sequence and state management.

P.S.
This comment is 100% human generated - though I'm curious to see if my writing style has been influenced by reading too much LLM output over the last few years

**Comment by projectgus** [comment]
> for OTA enabled devices, this is the one of the most important features,

We agree about this point, and certainly I'd be happy to see robustness improvements to the esp32 OTA system in MicroPython (not least because I know for a fact it used in some commercial deployments now.)

The question is about *how* to achieve them. As MicroPython maintainers we have to balance a number of factors in order to find the best outcome. It's not simple from our perspective.

> you can not rely solely on devs to implement it correctly. I wouldn't trust even myself for that. It is too risky to deviate from the official battle tested workflow recommended by Espressiff.

This is true even if we wrap the OTA API from Espressif, and is why we have `tests/ports/esp32/partition_ota.py`.

> I can make this feature optional via a flag in mpconfigport.h. Users who need the full OTA workflow can just enable it.
> What do you think?

This isn't a great solution unless we can enable it on most boards, because disabled-by-default features are hard to test and tend to bitrot. It'd also leave us in the situation of having an existing, documented, less robust OTA system (the current one) and a better one gated behind a compile flag. This itself is error-prone for developers.

BTW, if the 1.6KiB static memory usage I noted is unavoidable then we would have to leave it disabled on most boards. Do you have any idea where that comes from?

> Manually calling write -> verify -> set_boot is error-prone and requires reading the IDF documentation carefully

Not sure I agree. If we changed the set_boot API call to be `set_boot(verify=True)` then the default can easily do the same checks as `ota_end()`, and this would work even for existing MicroPython OTA update code which is a big win for existing users.

As far as I can see, the difference of "reading the IDF documentation carefully" comes down to manually erasing the region and then calling the partition write function which isn't particularly complex (and we could wrap this in a micropython-lib module to appear like a file object as well, if we had to). 

Am I missing something else?
